accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,871 @@
+            a set element type here, and a key/value pair here and rely on other
+            elements to disambiguate.)*
+
+        *   If `S` is `Null` and the spread operator is `...?` then the element
+            has set element type `Null`, map key type `Null` and map value type
+            `Null`.
+
+        *   If none of these cases match, it is an error.
+
+    *   If `P` is `Set<Ps>` then let `S` be the inferred type of `e1` in context
+        `Set<Ps>`:
+
+        *   If `S` is a non-`Null` subtype of `Iterable<Object>`, then the
+            inferred set element type of `element` is `T` where `T` is the type
+            such that `Iterable<T>` is a superinterface of `S` (the result of
+            constraint matching for `X` using the constraint `Iterable<X> <:
+            S`).
+
+        *   If `S` is `dynamic`, then the inferred set element type of `element`
+            is `dynamic`.
+
+        *   If `S` is `Null` and the spread operator is `...?`, then the set
+            element type is `Null`.
+
+        *   Otherwise it is an error.
+
+    *   If `P` is `Map<Pk, Pv>` then let `S` be the inferred type of `e1` in
+        context `P`:
+
+        *   If `S` is a non-`Null` subtype of `Map<Object, Object>`, then the
+            inferred key type of `element` is `K` and the inferred value type of
+            `element` is `V`, where `K` and `V` are the types such that `Map<K,
+            V>` is a superinterface of `S` (the result of constraint matching
+            for `X` and `Y` using the constraint `Map<X, Y> <: S`).
+
+        *   If `S` is `dynamic`, then the inferred key type of `element` is
+            `dynamic`, and the inferred value type of `element` is `dynamic`.
+
+        *   Otherwise it is an error.
+
+*   If `element` is an `ifElement` with one `element`, `p1`, and no "else"
+    element:
+
+    The condition is inferred with a context type of `bool`.
+
+    *   If the inferred set element type of `p1` is `S` then the inferred set
+        element type of `element` is `S`.
+
+    *   If the inferred key type of `p1` is `K` and the inferred value type of
+        `p1` is `V` then the inferred key and value types of `element` are `K`
+        and `V`.
+
+    *Note that both of the above cases can simultaneously apply because of
+    `dynamic` spreads.*
+
+*   If `element` is an `ifElement` with two `element`s, `e1` and `e2`:
+
+    The condition is inferred with a context type of `bool`.
+
+    It is a compile error if `e1` must be a set and `e2` must be a map or vice
+    versa. *In other words, you can't spread a map on one branch and a set on
+    the other. Since `dynamic` provides both set and map key/value types, a
+    `dynamic` in either branch does not run into this case.*
+
+    *   If the inferred set element type of `e1` is `S1` and the inferred set
+        element type of `e2` is `S2` then the inferred set element type of
+        `element` is the upper bound of `S1` and `S2`.
```

Comment:
So if one of them does not have a set element type, then the if-element doesn't have a set element type either. (LUB(type, undefined) = undefined).
That's actually reasonable.